### ISSUE TITLE
Default Content-Type for POST requests to be compatible with Rails params

### ISSUE THREAD

**Issue by jefflab** [puma/puma#695]
# Description

HTTP POST requests that are missing a "Content-Type" header get properly parsed in Unicorn. The params are not parsed in Puma. I think it makes sense to modify Puma to match Unicorn behavior to make migration easier.

# How to Repro

Set up a Rails controller to print the params:

```
def echo
  p params
end
```

**In Puma**

```
curl -d "foo=boar" http://localhost:5000/api/v1/echo
{"foo"=>"boar", "controller"=>"dummy", "action"=>"echo"}

curl -d "foo=boar" -H "Content-Type:" http://localhost:5000/api/v1/echo
{"controller"=>"dummy", "action"=>"echo"}
```

**In Unicorn**

```
curl -d "foo=boar" http://localhost:5000/api/v1/echo
{"foo"=>"boar", "controller"=>"dummy", "action"=>"echo"}

curl -d "foo=boar" -H "Content-Type:" http://localhost:5000/api/v1/echo
{"foo"=>"boar", "controller"=>"dummy", "action"=>"echo"}
```

# Questions

* Can we change the default behavior in Puma to match Unicorn for HTTP POST methods with missing Content-Type headers?
* Is the best way to override this with Rack middleware? If so, any pointers where to start (which config values to change, and where in the middleware pipeline it should go?)

**Comment by evanphx** [comment]
Hm. I suppose this would simple set the Content-Type header to if it's absent on a POST?

**Comment by jefflab** [comment]
Yep, exactly

**Comment by evanphx** [comment]
So I don't see unicorn setting the `Content-Type` anywhere. Can you paste in the `request.headers` from both?

**Comment by jefflab** [comment]
I haven't had a chance to get the `request.headers` for both. However, I'm pretty sure I remember the problem had something to do with this line of `binder.rb`

https://github.com/puma/puma/blob/master/lib/puma/binder.rb#L21-L24

```
# Rack blows up if this is an empty string, and Rack::Lint
# blows up if it's nil. So 'text/plain' seems like the most
# sensible default value.
"CONTENT_TYPE".freeze => "text/plain",
```

Because the `Content-Type` is being set to "text/plain", the POST params are not being converted to a hash in Rails. I think the default for a POST should be `application/x-www-form-urlencoded` in Puma.

My suspicion is that Unicorn doesn't set the `Content-Type` (as you discovered). As a result, Rails chooses to interpret POST params has a hash.

If this isn't a sufficient clue, and you want me to get the `request.headers`, please reply and I'll try to get that for you.

**Comment by evanphx** [comment]
Aaaaah. Ok! So basically if it's empty rails or something else infers it to the right type but if it's set to `text/plain` it doesn't. Got it!

**Comment by jefflab** [comment]
Great!

**Comment by evanphx** [comment]
Thit was fixed in a2b02294c8905.
